### ISSUE TITLE
Be able to customize cursor shape

### ISSUE THREAD

**Issue by yueyingjuesha** [holzschu/a-shell#210]
I notice the cursor shape is not normal, sometimes it’s a block and sometimes it’s insert, wondering if it’s possible to modify the cursor so it’s just using one shape?

**Comment by holzschu** [comment]
Hi, it's part of larger issue: there are two cursors, and I cannot control the second one. I could make it invisible, but then the selected area becomes invisible too. I've tried changing the color of the second cursor when the user selects something (and keeping it transparent otherwise) but the changes become effective later, which is very confusing.

This is the largest bug that I have no idea how to solve.

**Comment by personalizedrefrigerator** [comment]
## Changing the block cursor shape
This can be done using escape sequences (temporarily, until terminal restart)! 

In Python:
![C423CA80-79D7-4486-9B00-159976C0EEAA](https://user-images.githubusercontent.com/46334387/118549713-b345fc00-b710-11eb-98f9-82524b9e852b.jpeg)
![7BF50D90-47EE-4188-B00A-A87B0C8BC030](https://user-images.githubusercontent.com/46334387/118550037-1e8fce00-b711-11eb-9d19-2d0c1520de2d.jpeg)

See also https://github.com/holzschu/a-shell/issues/189#issuecomment-832079929

**Comment by personalizedrefrigerator** [comment]
One option for the second (vertical bar) cursor is to make the cursor color the background color, then change the background color on selection.

Some CSS file:
```css
/* In term_.document_ */
.hasSelection * {
  background-color: rgba(100,100,100,0.5) !important; 
  transition: background-color 0.1s ease;
}
```

```js
const termDoc = term_.document_;

termDoc.onselectionchang = () => {
   if (!termDoc.getSelection().isCollapsed) {
     termDoc.body.classList.add('hasSelection');
   } else {
     termDoc.body.classList.remove('hasSelection');
   }
};
```

**Comment by holzschu** [comment]
I have tried something similiar, but the changes occured one or two events after the selection, which was really confusing. However, I did not have the `transition` and the `!important`, so I'm going to try again.
In the first piece of code, shouldn't that be `caret-color` instead of `background-color`?

**Comment by personalizedrefrigerator** [comment]
> In the first piece of code, shouldn't that be caret-color instead of background-color?

No. The first one is mostly a hack -- changing the background color slightly makes the cursor/selection visible. See  https://github.com/mdn/browser-compat-data/blob/070b4d347d52abdd11cab88639977c225ec53b85/css/properties/caret-color.json#L39

### I would strongly suggest the second fix over the first.

**Comment by personalizedrefrigerator** [comment]
> In the first piece of code, shouldn't that be caret-color instead of background-color?

No. The first fix is really more of a workaround; it changes the background color of everything to make the selection visible (a filter: invert(1) might be better). See https://github.com/mdn/browser-compat-data/blob/070b4d347d52abdd11cab88639977c225ec53b85/css/properties/caret-color.json#L39.

I strongly recommend the second fix suggestion.

I can submit a pull request for this when I have something that works reasonably well!

**Comment by personalizedrefrigerator** [comment]
I plan to clean this up and submit a pull request, but this is what I'm using for now (run with `jsc --in-window` or `cat filename.js | jsi -w`).

```js
const termDoc = term_.document_;

termDoc.onselectionchange = () => {
   if (!termDoc.getSelection().isCollapsed) {
     termDoc.body.classList.add('hasSelection');
   } else {
     termDoc.body.classList.remove('hasSelection');
   }
};

const NEW_CSS_STYLES = `
body {
  transition: filter 0.3s ease;
}

body.hasSelection {
  filter: invert(0.2);
}

.hasSelection * {
    caret-color: rgba(255, 255, 0, 0.9) !important;
    transition: caret-color 0.3s ease;
}

* {
  caret-color: rgba(0,0,0,0.1) !important;
  transition: caret-color 0.3s ease;
}
`;

const newCssElem = document.createElement("style");
newCssElem.innerHTML = NEW_CSS_STYLES;
termDoc.body.appendChild(newCssElem);
```

**Comment by holzschu** [comment]
I think you can see what I meant by "changes occur after one or two events": first a gray selection, then it turns to yellow. And when you deselect, the cursor remains visible for a few seconds. We need to force the webview to redraw when the CSS changes.

**Comment by holzschu** [comment]
If we could act on the CSS and see the results instantaneously, having `caret-color: transparent` when there is no selection, and `caret-color: cursor-color` when selection is active would be enough.

**Comment by verhovsky** [comment]
When I start the app I see two cursors. After I run a command I see all 3

![image](https://github.com/holzschu/a-shell/assets/5687998/14baa766-76a2-4fce-8c3a-ed27c5c9518f)
